README.md
## nice_buttons
`nice_buttons` a beautiful, animated, and customizable 3D button for your flutter project.

### Platform Support

| Android | iOS | Web | MacOS | Linux | Windows |
| :-----: | :-: | :---: | :-----: | :-: | :---: |
| &#9989;  | &#9989; | &#9989; | &#9989;  | &#9989; | &#9989; |

![Sample Image](../main/example/nice_buttons.gif)

### Installation
Add it as a dependency in their `pubspec.yaml`:
```
dependencies:
  nice_buttons: ^0.0.3
```

### Customizations

| Attribute |    Type    |  Default  | Description  |
| :-------- | :--------: | :-------: | :-------------------------------------------------------------------|
| `startColor` |  `Color`  | `0xFF2ec8ff` | The gradient start color |
| `endColor` |  `Color`  | `0xFF529fff` | The gradient end color |
| `borderColor` |  `Color`  | `0xFF3489e9` | The color of the 3D border. |
| `progressColor` |  `Color`  | `white` | The color of circular progress indicator, defaults to white |
| `progressSize` |  `Double`  | `20` | The size of progress indicator circle, defaults to 20 |
| `gradientOrientation` |  `GradientOrientation`  | `Horizontal` | Orientation of the gradient defaults to Horizontal |
| `borderThickness` |  `Double`  | `5` | The thickness of the 3D border, defaults to 5 |
| `height` |  `Double`  | `60` | Height of the button, defaults to 60 |
| `width` |  `Double`  | `200` | Width of the button defaults to 200, its useless if the stretch property is set to true |
| `stretch` |  `bool`  | `true` | Whether to occupy the full available space in the parent, defaults to true |
| `borderRadius` |  `Double`  | `20` | The border radius of the button, defaults to 20 |
| `progress` |  `bool`  | `false` | Whether the progress indicator is required or not, defaults to false |
| `disabled` |  `bool`  | `false` | Disables the button, defaults to false |
| `onTap` |  `Function`  |  | Button press handler, required* |
| `child` |  `Widget`  |  | Inner content for the button, required* |

### Usage
```
...
import 'package:nice_buttons/nice_buttons.dart';
...
```

```
NiceButtons(
  stretch: true,
  gradientOrientation: GradientOrientation.Horizontal,
  onTap: (finish) {
    print('On tap called');
  },
  child: Text(
    'Full Width',
    style: TextStyle(color: Colors.white, fontSize: 18),
  ),
);
```
Check example folder for more examples.


CHANGELOG.md
## [0.0.2]

- Added support for new flutter SDKs.

## [0.0.2]

- Updated the read me and package description.

## [0.0.1]

- Initial release


example/lib/main.dart
import 'package:flutter/material.dart';
import 'package:nice_buttons/nice_buttons.dart';
import 'dart:async';

void main() {
  runApp(MyApp());
}

class MyApp extends StatelessWidget {
  // This widget is the root of your application.
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        // This is the theme of your application.
        //
        // Try running your application with "flutter run". You'll see the
        // application has a blue toolbar. Then, without quitting the app, try
        // changing the primarySwatch below to Colors.green and then invoke
        // "hot reload" (press "r" in the console where you ran "flutter run",
        // or simply save your changes to "hot reload" in a Flutter IDE).
        // Notice that the counter didn't reset back to zero; the application
        // is not restarted.
        primarySwatch: Colors.blue,
      ),
      home: MyHomePage(title: 'Nice Buttons Demo'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  MyHomePage({Key? key, this.title}) : super(key: key);

  // This widget is the home page of your application. It is stateful, meaning
  // that it has a State object (defined below) that contains fields that affect
  // how it looks.

  // This class is the configuration for the state. It holds the values (in this
  // case the title) provided by the parent (in this case the App widget) and
  // used by the build method of the State. Fields in a Widget subclass are
  // always marked "final".

  final String? title;

  @override
  _MyHomePageState createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  @override
  Widget build(BuildContext context) {
    // This method is rerun every time setState is called, for instance as done
    // by the _incrementCounter method above.
    //
    // The Flutter framework has been optimized to make rerunning build methods
    // fast, so that you can just rebuild anything that needs updating rather
    // than having to individually change instances of widgets.
    return Scaffold(
      appBar: AppBar(
        // Here we take the value from the MyHomePage object that was created by
        // the App.build method, and use it to set our appbar title.
        title: Text(widget.title!),
      ),
      body: Center(
        // Center is a layout widget. It takes a single child and positions it
        // in the middle of the parent.
        child: Column(
          // Column is also a layout widget. It takes a list of children and
          // arranges them vertically. By default, it sizes itself to fit its
          // children horizontally, and tries to be as tall as its parent.
          //
          // Invoke "debug painting" (press "p" in the console, choose the
          // "Toggle Debug Paint" action from the Flutter Inspector in Android
          // Studio, or the "Toggle Debug Paint" command in Visual Studio Code)
          // to see the wireframe for each widget.
          //
          // Column has various properties to control how it sizes itself and
          // how it positions its children. Here we use mainAxisAlignment to
          // center the children vertically; the main axis here is the vertical
          // axis because Columns are vertical (the cross axis would be
          // horizontal).
          mainAxisAlignment: MainAxisAlignment.center,
          children: <Widget>[
            NiceButtons(
              stretch: false,
              progress: true,
              gradientOrientation: GradientOrientation.Horizontal,
              onTap: (finish) {
                print('On tap called');
                Timer(Duration(seconds: 5), () {
                  finish();
                });
              },
              child: Text(
                'With Progress',
                style: TextStyle(color: Colors.white, fontSize: 18),
              ),
            ),
            Container(width: double.infinity, height: 30),
            NiceButtons(
              stretch: false,
              gradientOrientation: GradientOrientation.Horizontal,
              onTap: (finish) {
                print('On tap called');
              },
              child: Text(
                'Nice Buttons',
                style: TextStyle(color: Colors.white, fontSize: 18),
              ),
            ),
            Container(width: double.infinity, height: 30),
            NiceButtons(
              stretch: true,
              gradientOrientation: GradientOrientation.Horizontal,
              onTap: (finish) {
                print('On tap called');
              },
              child: Text(
                'Full Width',
                style: TextStyle(color: Colors.white, fontSize: 18),
              ),
            ),
            Container(width: double.infinity, height: 30),
            NiceButtons(
              stretch: false,
              borderRadius: 30,
              gradientOrientation: GradientOrientation.Horizontal,
              onTap: (finish) {
                print('On tap called');
              },
              child: Text(
                '3D Pill',
                style: TextStyle(color: Colors.white, fontSize: 18),
              ),
            ),
            Container(width: double.infinity, height: 30),
            NiceButtons(
              stretch: false,
              gradientOrientation: GradientOrientation.Vertical,
              onTap: (finish) {
                print('On tap called');
              },
              child: Text(
                'Vertical Gradient',
                style: TextStyle(color: Colors.white, fontSize: 18),
              ),
            ),
          ],
        ),
      ),
    );
  }
}


lib/nice_buttons copy.dart
library nice_buttons;

import 'package:flutter/material.dart';

class NiceButtonsStyle {
  final Color topColor;
  final Color backColor;
  final BorderRadius borderRadius;
  final double z;
  final double tappedZ;
  const NiceButtonsStyle(
      {this.topColor = const Color(0xFF45484c),
      this.backColor = const Color(0xFF191a1c),
      this.borderRadius = const BorderRadius.all(Radius.circular(7.0)),
      this.z = 8.0,
      this.tappedZ = 3.0});

  static const RED = const NiceButtonsStyle(
      topColor: const Color(0xFFc62f2f), backColor: const Color(0xFF922525));
  static const BLUE = const NiceButtonsStyle(
      topColor: const Color(0xFF25a09c), backColor: const Color(0xFF197572));
  static const WHITE = const NiceButtonsStyle(
      topColor: const Color(0xFFffffff), backColor: const Color(0xFFCFD8DC));
}

class NiceButtons extends StatefulWidget {
  final VoidCallback onPressed;
  final Widget child;
  final NiceButtonsStyle style;
  final double width;
  final double height;

  NiceButtons(
      {@required this.onPressed,
      @required this.child,
      this.style = NiceButtonsStyle.WHITE,
      this.width = 100.0,
      this.height = 90.0});

  @override
  State<StatefulWidget> createState() => NiceButtonsState();
}

class NiceButtonsState extends State<NiceButtons> {
  bool isTapped = false;

  Widget _buildBackLayout() {
    return Padding(
      padding: EdgeInsets.only(top: widget.style.z),
      child: DecoratedBox(
        position: DecorationPosition.background,
        decoration: BoxDecoration(
            borderRadius: widget.style.borderRadius,
            boxShadow: [BoxShadow(color: widget.style.backColor)]),
        child: ConstrainedBox(
          constraints: BoxConstraints.expand(
              width: widget.width, height: widget.height - widget.style.z),
        ),
      ),
    );
  }

  Widget _buildTopLayout() {
    return Padding(
      padding: EdgeInsets.only(
          top: isTapped ? widget.style.z - widget.style.tappedZ : 0.0),
      child: DecoratedBox(
        position: DecorationPosition.background,
        decoration: BoxDecoration(
            borderRadius: widget.style.borderRadius,
            boxShadow: [BoxShadow(color: widget.style.topColor)]),
        child: ConstrainedBox(
          constraints: BoxConstraints.expand(
              width: widget.width, height: widget.height - widget.style.z),
          child: Container(
            padding: EdgeInsets.zero,
            alignment: Alignment.center,
            child: widget.child,
          ),
        ),
      ),
    );
  }

  @override
  Widget build(BuildContext context) {
    return GestureDetector(
      child: Stack(
        alignment: AlignmentDirectional.topCenter,
        children: <Widget>[_buildBackLayout(), _buildTopLayout()],
      ),
      onTapDown: (TapDownDetails event) {
        setState(() {
          isTapped = true;
        });
      },
      onTapCancel: () {
        setState(() {
          isTapped = false;
        });
      },
      onTapUp: (TapUpDetails event) {
        setState(() {
          isTapped = false;
        });
        widget.onPressed();
      },
    );
  }
}


lib/nice_buttons.dart
library nice_buttons;

import 'package:flutter/material.dart';

/// Gradient settings enum, this can be used to change the direction of the button
/// color gradients.
enum GradientOrientation {
  Vertical,
  Horizontal,
}

class NiceButtons extends StatefulWidget {
  /// The gradient start color from left when gradientOrientation is horizontal,
  /// from top when orientation is vertical.
  final Color startColor;

  /// The gradient end color to right when orientation is horizontal and to
  /// bottom when orientation is vertical.
  final Color endColor;

  /// The color of the 3D border of the button.
  final Color borderColor;

  /// The color of circular progress indicator, defaults to white.
  final Color progressColor;

  /// The size of progress indicator circle, defaults to 20.
  final double progressSize;

  /// Orientation of the gradient defaults to Horizontal.
  final GradientOrientation gradientOrientation;

  /// The thickness of the 3D border, defaults to 5.
  final double borderThickness;

  /// Height of the button, defaults to 60.
  final double height;

  /// The border radius for the button, defaults to 20.
  final double borderRadius;

  /// Whether to occupy the full available space in the parent, defaults to true.
  final bool stretch;

  /// Width of the button defaults to 200, it's useless if the stretch property
  /// is set to true.
  final double width;

  /// Whether the progress indicator is required or not, defaults to false.
  final bool progress;

  /// Disables the button, defaults to false.
  final bool disabled;

  /// Button on press handler, required.
  final Function onTap;

  /// Child widget that will be wrapped inside the button.
  final Widget child;

  /// The calculated border radius for the buttons.
  final BorderRadius br;

  /// The calculated width of the button, sets to parent width when
  /// stretch property is set to true.
  final double calculatedWidth;

  NiceButtons({
    required this.onTap,
    required this.child,
    this.startColor = const Color(0xFF2ec8ff),
    this.endColor = const Color(0xFF529fff),
    this.borderColor = const Color(0xFF3489e9),
    this.progressColor = Colors.white,
    this.progressSize = 20,
    this.borderRadius = 20,
    this.borderThickness = 5,
    this.height = 60,
    this.width = 200,
    this.gradientOrientation = GradientOrientation.Vertical,
    this.stretch = true,
    this.progress = false,
    this.disabled = false,
  })  : this.br = BorderRadius.all(Radius.circular((borderRadius))),
        this.calculatedWidth = stretch ? double.infinity : width;

  @override
  _NiceButtonsState createState() => _NiceButtonsState(borderThickness);
}

/// Button state to keep track of press and progress animations.
class _NiceButtonsState extends State<NiceButtons>
    with TickerProviderStateMixin {
  double _borderThickness = 5;
  double _moveMargin = 0.0;
  double _progressWidth = 0.0;
  bool _showProgress = false;
  bool _tapped = false;
  bool _processing = false;
  int _progressBarMillis = 2500;

  _NiceButtonsState(double borderThickness) {
    this._borderThickness = borderThickness;
  }

  /// The back view of the button that gives the feel of 3D border.
  Widget _buildBackLayout() {
    return Container(
      padding: EdgeInsets.only(top: _borderThickness),
      width: double.infinity,
      height: double.infinity,
      child: DecoratedBox(
        position: DecorationPosition.background,
        decoration: BoxDecoration(
          borderRadius: widget.br,
          color: widget.borderColor,
        ),
        child: ConstrainedBox(
          constraints: BoxConstraints.expand(
              width: double.infinity, height: double.infinity),
        ),
      ),
    );
  }

  /// The front view of the button, that has the user child widget content and
  /// progress widget to show progress animation.
  Widget _buildFrontLayout() {
    return AnimatedContainer(
      onEnd: () {
        setState(() {
          _moveMargin = 0;
          if (widget.progress && !_showProgress && _tapped) {
            _showProgress = true;
            _progressWidth = double.infinity;
            _processing = true;
            _progressBarMillis = 2500;
          }
          _tapped = false;
        });
      },
      margin: EdgeInsets.only(top: _moveMargin),
      duration: Duration(milliseconds: 150),
      curve: Curves.easeInOut,
      decoration: BoxDecoration(
        borderRadius: widget.br,
        gradient: LinearGradient(
          begin: widget.gradientOrientation == GradientOrientation.Vertical
              ? Alignment.topCenter
              : Alignment.centerLeft,
          end: widget.gradientOrientation == GradientOrientation.Vertical
              ? Alignment.bottomCenter
              : Alignment.centerRight,
          colors: [
            widget.startColor,
            widget.endColor,
          ],
        ),
      ),
      child: ConstrainedBox(
        constraints: BoxConstraints.expand(
            width: double.infinity, height: widget.height - _borderThickness),
        child: ClipRRect(
          borderRadius: widget.br,
          child: Stack(
            children: <Widget>[
              _buildProgressBar(),
              if (_showProgress) _buildProgressCircle(),
              if (!_showProgress) _buildUserChild(),
            ],
          ),
        ),
      ),
    );
  }

  /// The progress widget.
  Widget _buildProgressBar() {
    return AnimatedSize(
      duration: Duration(milliseconds: _progressBarMillis),
      curve: Curves.fastOutSlowIn,
      child: Container(
        width: _progressWidth,
        height: double.infinity,
        color: new Color.fromARGB(60, 255, 255, 255),
      ),
    );
  }

  /// The progress circle widget.
  Widget _buildProgressCircle() {
    return Container(
      width: double.infinity,
      height: double.infinity,
      child: Align(
        alignment: Alignment.center,
        child: SizedBox(
          width: widget.progressSize,
          height: widget.progressSize,
          child: CircularProgressIndicator(
              valueColor:
                  new AlwaysStoppedAnimation<Color>(widget.progressColor)),
        ),
      ),
    );
  }

  /// User provided child widget.
  Widget _buildUserChild() {
    return Align(
      alignment: Alignment.center,
      //child: widget.child != null ? widget.child : Text(''),
      child: widget.child,
    );
  }

  /// The button press handler for the gesture detector. This will be triggered
  /// when button is pressed. Calls the user provided onTap function.
  void _onTap() {
    setState(() {
      _moveMargin = _borderThickness;
      _tapped = true;
    });
    widget.onTap(_finish);
  }

  /// The finish method that will reset the state of the animation, user
  /// has to call this explicitly from the parent.
  void _finish() {
    setState(() {
      _showProgress = false;
      _progressWidth = 0;
      _processing = false;
      _progressBarMillis = 1;
    });
  }

  /// The cumulative button widget that builds the front and back layout of the
  /// button.
  @override
  Widget build(BuildContext context) {
    return GestureDetector(
      onTap: widget.disabled || _processing ? null : _onTap,
      child: Container(
        child: ConstrainedBox(
          constraints: BoxConstraints.expand(
              width: widget.calculatedWidth, height: widget.height),
          child: Stack(
            children: <Widget>[
              _buildBackLayout(),
              _buildFrontLayout(),
            ],
          ),
        ),
      ),
    );
  }
}
